### PR TITLE
Make `sf::Clock` pausable

### PR DESCRIPTION
**Pull request by MetGang**
* [X] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before?
* [X] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
* [X] Have you provided some example/test code for your changes?
* [ ] If you have additional steps which need to be performed list them as tasks!

## Description

Ability to pause `sf::Clock` was discussed and proposed on SFML forum many times - yet it was not implemented (and there is no plans to do so) in SFML 2.x. With this PR this goal is achieved (and some more) with SFML 3 release in mind.
This is very light PR that may cause more changes to come in the future.

## Tasks

* [ ] Tested on Linux
* [X] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

Internal repository tests have been updated to cover changes from this PR.
*Passed on my machine.*

### CODE REVIEW

File: include/SFML/System/Clock.hpp
Comment (ChrisThrasher) [inline include/SFML/System/Clock.hpp:121]: Is this equivalent to `clock.getElapsedTime() > time`?

File: include/SFML/System/Time.hpp
Comment (ChrisThrasher) [inline include/SFML/System/Time.hpp:91]: I believe you're the first person to start (rightfully) decorating functions as `noexcept` where appropriate. This PR would be the first to introduce such changes. @eXpl0it3r @Bromeon @vittorioromeo is this a pattern we should start applying to everything which cannot throw?

File: test/System/Clock.cpp
Comment (ChrisThrasher) [inline test/System/Clock.cpp:31]: Should we check that `isRunning` is false after construction?

File: test/System/Clock.cpp
Comment (ChrisThrasher) [inline test/System/Clock.cpp:38]: Should we check that `isPaused` is false after construction?

**Comment by Bromeon** [comment]
Regarding `[[nodiscard]]` and `noexcept`:
[Forum thread](https://en.sfml-dev.org/forums/index.php?topic=28340.0)
[GitHub comment about `[[nodiscard]]`](https://github.com/SFML/SFML/pull/1883#issuecomment-989583140)

Would be nice if we didn't have to reiterate the discussion for every pull request 🙂

File: include/SFML/System/Time.hpp
Comment (Bromeon) [reply include/SFML/System/Time.hpp:91]: see main thread

**Review by Bromeon — COMMENTED** [review]
Thanks for the contribution!

This changes a bit many things at once, e.g. the `[[nodiscard]]`/`noexcept` is highly controversial and shouldn't be mixed in a functionality-PR, if it changes a lot of code beyond the additions. See my message above for rationale.

Also, it's not really clear whether stop-watches are something we need/want in SFML, and what the design would be. For example, the current approach makes `sf::Clock` bigger, even if the pause functionality is not used.

File: test/System/Clock.cpp
Comment (MetGang) [reply test/System/Clock.cpp:31]: That's a good point since clock by default is running.

File: test/System/Clock.cpp
Comment (MetGang) [reply test/System/Clock.cpp:38]: According to my previous response `!isPaused` is more appropriate.

File: include/SFML/System/Clock.hpp
Comment (MetGang) [reply include/SFML/System/Clock.hpp:121]: It is equivalent. Motivation was just to have dedicated method to perform these checks. I understand some may find this addition redundant. Also I can see your point with this method being less versetile than comparision operators which sf::Time has plenty.

File: include/SFML/System/Clock.hpp
Comment (MetGang) [reply include/SFML/System/Clock.hpp:151]: I read posts about `[[nodiscard]]`/`noexcept` and I understand that this PR may push this concept too much when its main goal is ability to pause a clock. Nonetheless, sooner or later these changes **may** be necessary but now for I'll reiterate changes for more consistency.

As for the point with size increase there was idea to seperate this feature and make `sf::PausableClock` along with unchanged `sf::Clock`. Then, user wouldn't have to pay for pause feature.

`toggle()` was motivated with simplifying code like below but since `setX(bool)` is more used in library and is more versatile I understand your point. It can substituted with `clock.setRunning(!clock.isRunning())` though.
```cpp
if (clock.isRunning())
{
    clock.pause();
}
else
{
    clock.start();
}
```

With addition of `setRunning(bool)` I'd drop `isPaused()` and keep `isRunning()` for consistency but keep `start()` and `pause()`.

I'll wait for responses and more feedback then commit accordingly ;)

**Comment by vittorioromeo** [comment]
> For example, the current approach makes `sf::Clock` bigger, even if the pause functionality is not used.

@Bromeon: This is technically true, but it really is insignificant. :)

---

@MetGang: I would simplify the entire PR by a lot. Drop all the unnecessary changes (e.g. `noexcept` and `[[nodiscard]]`), replace the multiple functions with `setRunning`, remove `hasTimePassed`. It will be easier for us to review it and approve it. :)

File: src/SFML/System/Clock.cpp
Comment (kimci86) [inline src/SFML/System/Clock.cpp:56]: This implicitly requires `ClockImpl` to give time in microseconds, which might not be the case.

File: src/SFML/System/Clock.cpp
Comment (kimci86) [inline src/SFML/System/Clock.cpp:122]: Why not `m_pausePoint == ClockImpl::time_point{}`?

File: src/SFML/System/Clock.cpp
Comment (Bromeon) [reply src/SFML/System/Clock.cpp:47]: As mentioned by ChrisThrasher and vittorio already, this method should go entirely.

File: src/SFML/System/Clock.cpp
Comment (Bromeon) [inline src/SFML/System/Clock.cpp:100]: Please use `type()` constructor syntax for consistency

File: src/SFML/System/Clock.cpp
Comment (MetGang) [reply src/SFML/System/Clock.cpp]: Addressed in new commits.

File: src/SFML/System/Clock.cpp
Comment (MetGang) [reply src/SFML/System/Clock.cpp]: Addressed in new commits.

File: src/SFML/System/Clock.cpp
Comment (MetGang) [reply src/SFML/System/Clock.cpp]: Addressed in new commits.

File: src/SFML/System/Clock.cpp
Comment (MetGang) [reply src/SFML/System/Clock.cpp:121]: `m_pausePoint == ClockImpl::time_point()` is also a valid (and maybe simpler) way but I'd personally stick to existing one since it reflects intentions (`m_pausePoint` must be zero to identify that clock is running) better. I'm open to hear others' opinions though.

File: src/SFML/System/Clock.cpp
Comment (MetGang) [reply src/SFML/System/Clock.cpp]: Addressed in new commits.

File: include/SFML/System/Clock.hpp
Comment (ananace) [inline include/SFML/System/Clock.hpp:195]: Typo, should be `std::chrono`

Comment (MetGang) [reply include/SFML/System/Clock.hpp:195]: I misstyped `std::chrono` as `std::chorno`

Comment (DelinWorks) [reply include/SFML/System/Clock.hpp]: What they look soo the same 😂

File: include/SFML/System/Time.hpp
Comment (ChrisThrasher) [inline include/SFML/System/Time.hpp:4]: Did you mean to change the copyright year on a file you didn't modify?

File: include/SFML/System/Time.inl
Comment (ChrisThrasher) [inline include/SFML/System/Time.inl:4]: Did you mean to change this copyright year?

File: src/SFML/System/Time.cpp
Comment (ChrisThrasher) [inline src/SFML/System/Time.cpp:33]: Did you mean to modify whitespace in a file you otherwise didn't change?

File: src/SFML/System/Clock.cpp
Comment (ChrisThrasher) [inline src/SFML/System/Clock.cpp:101]: Could you remove a few of these extra empty newlines to better match the style of existing code? Most SFML code puts expressions on consecutive lines instead of double-spacing everything.

File: src/SFML/System/Clock.cpp
Comment (ChrisThrasher) [inline src/SFML/System/Clock.cpp:114]: See previous commenting about double-spacing all code.

**Review by Bromeon — COMMENTED** [review]
Having 6 functions with overlapping semantics, yet completely different terminology is confusing:
* (constructor)
* setRunning
* start
* pause
* reset
* restart

I also believe we don't have to provide every possible combination of operations "for convenience", and we might want to split functionalty. Maybe `start`/`stop` are still justified, as most of the cases, the logic around will know whether the stopwatch should be running or not, and thus `setRunning(bool)` might not be necessary.

What if we keep the API simple:
```cpp
class Clock {
public:
    // Running state
    void start();
    void stop(); // formerly pause()
    bool isRunning() const;

    // Elapsed time state
    void reset();
    Time getElapsedTime() const;
};
```

One option would be to just keep the two states completely orthogonal, so `reset()` would not affect running/stopped state. But in practice, one is likely to either stop the clock or restart it, so maybe `reset()` can have a fixed meaning.

I'm not sure if we need the methods who act as both setters and getters simultaneously (current `restart()`) -- while convenient, it introduces some redundancy and makes the API surface broader. So far there was only one state (elapsed time), so it was manageable, but now we additionally introduce running/stopped, so we might focus on making sure the user always knows which state he is manipulating.

File: include/SFML/System/Clock.hpp
Comment (MetGang) [reply include/SFML/System/Clock.hpp:119]: Thanks for pointing out docs error.

File: src/SFML/System/Clock.cpp
Comment (MetGang) [reply src/SFML/System/Clock.cpp:101]: Seems reasonable, gonna address that.

File: include/SFML/System/Time.hpp
Comment (MetGang) [reply include/SFML/System/Time.hpp:4]: Actually Time files were changed too, then changed again and there is some leftover after those changes. I think I should revert all changes then.

File: include/SFML/System/Clock.hpp
Comment (MetGang) [reply include/SFML/System/Clock.hpp]: Regarding API changes. Changing `pause()` into `stop()` is okay. Taking return values from `restart/reset()` as form of simplifying API since we manage two states now - okay. Though I would not consider deleting `restart()` method.

**Comment by ChrisThrasher** [comment]
This PR has become significantly diverged from the `master` branch. Can you rebase and squash on `master`?

**Comment by ChrisThrasher** [comment]
Can you squash this into some smaller number of commits and remove the merge commits you added? That'll have to happen before it can get merged.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
@MetGang Are you still interested in maintaining this PR?

**Comment by MetGang** [comment]
@ChrisThrasher Sure, it seems that clang-format changes from master result in conflicts with my changes. Should be easy to fix when I find some time.

**Comment by MetGang** [comment]
Yep... I metged a wrong thing here again ^^"

**Comment by ChrisThrasher** [comment]
I updated this to fix all the merge conflicts. This idea is growing on me.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Review by ChrisThrasher — APPROVED** [review]

**Comment by ChrisThrasher** [comment]
I renamed `pause()` to `stop()` and removed `setRunning()` since I'd rather users call `start()` and `stop()` explicitly. I still do like `reset()` and `restart()` returning a `Time` so I hope that can stay. I find that really convenient. I hope we can make a decision on this since soon.

SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by estebanc08** [comment]
I am currently using a clock object to synchronize the notes being played on piano to a midi file that reads in music as timed ticks. Im wanting to add a function to pause the notes and the music and therefore pause the clock so the notes playing can stay synchronized with the clock and a pause function is really the only way I see the issue being resolved.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by Bambo-Borris** [comment]
On a project of my own I've encountered a need for this, and I've also helped users in the SFML Discord recently who could have benefitted from this as a feature. I think it's a good win in terms of improving a very commonly used utility class from SFML.

**Review by Bambo-Borris — APPROVED** [review]

**Comment by ChrisThrasher** [comment]
Rebased on `master` and fixed conflicts

SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Review by binary1248 — APPROVED** [review]

**Comment by kimci86** [comment]
One small inconsistency is that a default-constructed clock is running but a clock after `reset` is not running.
One could expect `reset` to put the clock in the state of a default-constructed clock.

Looking online for other similar interfaces, I see a clock (or rather stopwatch) is not running after construction there:
- .NET: https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch
- Apache Commons: https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/time/StopWatch.html
- Dart: https://api.dart.dev/stable/3.0.2/dart-core/Stopwatch-class.html

Maybe a default-constructed clock should be stopped?
Because creating a running clock can be convenient, we could add a parameter to the clock constructor or maybe a factory function? This is what Apache commons and Guava do.
- Guava: https://guava.dev/releases/18.0/api/docs/com/google/common/base/Stopwatch.html

**Comment by ChrisThrasher** [comment]
Good point on a clock maybe being stopped by default. Ideally we'd rename it to sf::Stopwatch and thus better justify some radical changes. However if we keep the same name and keep the default constructor, I'm aprehensive about a change that will break so many existing programs at runtime.

**Comment by Bromeon** [comment]
We could also consider named constructor idiom:
```cpp
auto s = sf::Clock::stopped();
auto r = sf::Clock::running();
```

Do we need a default constructor? It may be convenient, but it's not obvious what it does, as the discussion here shows.

A problem with the above proposal is that static methods can also be called on an instance (another huge C++ mistake), i.e.
```cpp
sf::Clock c = ...;

sf::Clock d = c.stopped(); // does not touch c
```

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: test/System/Clock.test.cpp
Comment (eXpl0it3r) [inline test/System/Clock.test.cpp:43]: There's no test case for the behavior of:

- Stop a running clock
- Take measure of the elapsed time
- Start a stopped clock
- Elapsed time isn't reset

**Review by eXpl0it3r — APPROVED** [review]
I'm not entirely convinced of the API, as I feel `start()`, `restart()` and `reset()` aren't exactly obvious in their operations. Why does start not reset the elapsed time, yet restart does? Why does reset leave the clock in a stopped state, when the default construction is a started state?

And I think the whole change really makes a strong argument for a different name, for which we have [a discussion thread on the forum](https://en.sfml-dev.org/forums/index.php?topic=28969.0) already.

Why an approval then? I'd like to see progress made and maybe it's not the final state of the whole feature, but we're also done with SFML 3, so lets get it in and maybe think about the fine tuning of the API along the way.  
It's a bit "radical" to how we've done stuff otherwise in SFML, but happy to try something a bit different.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
Thanks everyone for their feedback on this new feature! Let's continue the API discussion in #https://github.com/SFML/SFML/issues/2570.

**Pull request merged**
